packages/devtools-source-map/src/utils/fetchSourceMap.js

```diff
@@ -0,0 +1,102 @@
+  // wire.
+  if (sourceMap.hasContentsOfAllSources()) {
+    return;
+  }
+
+  // If it's already a URL, just leave it alone.
+  if (!path.isURL(sourceMap.sourceRoot)) {
+    // In the odd case where the sourceMap is a data: URL and it does
+    // not contain the full sources, fall back to using the source's
+    // URL, if possible.
+    let parsedSourceMapURL = new URL(absSourceMapURL);
+    if (parsedSourceMapURL.protocol === "data:" && source.url) {
+      parsedSourceMapURL = new URL(source.url);
+    }
+
+    parsedSourceMapURL.pathname = path.dirname(parsedSourceMapURL.pathname);
+    sourceMap.sourceRoot = new URL(sourceMap.sourceRoot || "",
+                                   parsedSourceMapURL).toString();
+  }
+  return sourceMap.sourceRoot;
+}
+
+function _resolveSourceMapURL(source: Source) {
+  const { url = "", sourceMapURL = "" } = source;
+  if (path.isURL(sourceMapURL) || url == "") {
+    // If it's already a full URL or the source doesn't have a URL,
+    // don't resolve anything.
+    return sourceMapURL;
+  } else if (path.isAbsolute(sourceMapURL)) {
+    // If it's an absolute path, it should be resolved relative to the
+    // host of the source.
+    const { protocol = "", host = "" } = new URL(url);
+    return `${protocol}//${host}${sourceMapURL}`;
+  }
+  // Otherwise, it's a relative path and should be resolved relative
+  // to the source.
+  return `${path.dirname(url)}/${sourceMapURL}`;
+}
+
+
+async function _resolveAndFetch(generatedSource: Source) : SourceMapConsumer {
+  // Fetch the sourcemap over the network and create it.
+  const sourceMapURL = _resolveSourceMapURL(generatedSource);
+  const fetched = await networkRequest(
+    sourceMapURL, { loadFromCache: false }
+  );
+
+  // Create the source map and fix it up.
+  let map = new SourceMapConsumer(fetched.content);
+  if (generatedSource.isWasm) {
+    map = new WasmRemap(map);
+  }
+
+  _setSourceMapRoot(map, sourceMapURL, generatedSource);
+  return map;
+}
+
+function fetchSourceMap(generatedSource: Source) {
+  const existingRequest = getSourceMap(generatedSource.id);
+  if (existingRequest) {
+    // If it has already been requested, return the request. Make sure
+    // to do this even if sourcemapping is turned off, because
+    // pretty-printing uses sourcemaps.
+    //
+    // An important behavior here is that if it's in the middle of
+    // requesting it, all subsequent calls will block on the initial
+    // request.
+    return existingRequest;
+  } else if (!generatedSource.sourceMapURL) {
+    return Promise.resolve(null);
+  }
+
+  // Fire off the request, set it in the cache, and return it.
+  const req = _resolveAndFetch(generatedSource);
+  // Make sure the cached promise does not reject, because we only
+  // want to report the error once.
+  setSourceMap(generatedSource.id, req.catch(() => null));
+  return req;
+}
+
+module.exports = { fetchSourceMap }
```

Comment:
I like that in the case of fetch, it's clear that we only expose one function. Previously, we were relying on `_` convention, but it wasn't clear if other APIs were inappropriately reaching in and making calls to the other fetching utils